### PR TITLE
SLING-10164 : Add basic metrics to SlingAuthenticator 

### PR DESCRIPTION
**Pull request by anchela**
(initial proposal with minimal testing)

### CODE REVIEW

**Comment by anchela** [comment]
@cziegeler , compared to the original draft attached in jira, i slightly modified the metric names (dropping org.apache) and also added some very limited testing. as mentioned in jira, i would appreciate if someone would carefully review (both from an authenticator and metric point of view).

File: src/main/java/org/apache/sling/auth/core/impl/SlingAuthenticator.java [hunk 1]
```diff
@@ -457,6 +465,13 @@ private void deactivate(final BundleContext bundleContext) {
             webConsolePlugin.unregister();
             webConsolePlugin = null;
         }
+
+        if (metricsService != null) {
```
Comment (bdelacretaz) [inline src/main/java/org/apache/sling/auth/core/impl/SlingAuthenticator.java:469]: Why not just `metricsService = null` ? Not sure why you'd need the test, as no method is called on metricsService

File: src/main/java/org/apache/sling/auth/core/impl/SlingAuthenticator.java [hunk 2]
```diff
@@ -492,15 +507,17 @@ public boolean handleSecurity(HttpServletRequest request,
...
         if (process && expectAuthenticationHandler(request)) {
             log.warn("handleSecurity: AuthenticationHandler did not block request; access denied");
             request.removeAttribute(AuthenticationHandler.FAILURE_REASON);
             request.removeAttribute(AuthenticationHandler.FAILURE_REASON_CODE);
             AuthUtil.sendInvalid(request, response);
-            return false;
+            process = false;
         }
-
+        ctx.stop();
```
Comment (bdelacretaz) [inline src/main/java/org/apache/sling/auth/core/impl/SlingAuthenticator.java:519]: Without having looked at whether that's really needed, putting those stop + authenticateCompleted statements in a `finally` block sounds safer to me.

File: src/main/java/org/apache/sling/auth/core/impl/SlingAuthenticator.java [hunk 3]
```diff
@@ -492,15 +507,17 @@ public boolean handleSecurity(HttpServletRequest request,
...
+        metrics.authenticateCompleted(process);
```
Comment (bdelacretaz) [inline src/main/java/org/apache/sling/auth/core/impl/SlingAuthenticator.java:520]: As per this method's javadoc, process==false means authentication failed. But passing false to authenticateCompleted means there's no failure, isn't that reversed?

File: src/main/java/org/apache/sling/auth/core/impl/SlingAuthenticationMetrics.java [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+        authenticateSuccess = metricsService.meter("sling.auth.core.authenticate.success");
+        authenticateFailed = metricsService.meter("sling.auth.core.authenticate.failed");
+    }
+
+    @NotNull
+    Timer.Context authenticationTimerContext() {
+        return authenticateTimer.time();
+    }
+
+    void authenticateCompleted(boolean failed) {
```
Comment (bdelacretaz) [inline src/main/java/org/apache/sling/auth/core/impl/SlingAuthenticationMetrics.java:41]: taking a "failed" argument sounds counterintuitive to me, see also my comment below where this is called.

File: src/test/java/org/apache/sling/auth/core/impl/SlingAuthenticatorOsgiTest.java [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+    private final SlingAuthenticator authenticator = new SlingAuthenticator();
+
+    @Before
+    public void before() throws Exception {
+        ResourceResolver rr = mock(ResourceResolver.class);
+        ResourceResolverFactory resourceResolverFactory = mock(ResourceResolverFactory.class);
+        when(resourceResolverFactory.getResourceResolver(any(AuthenticationInfo.class))).thenReturn(rr);
+
+        when(timer.time()).thenReturn(ctx);
+        when(metricsService.meter(anyString())).thenReturn(meter);
```
Comment (bdelacretaz) [inline src/test/java/org/apache/sling/auth/core/impl/SlingAuthenticatorOsgiTest.java:63]: using the same Meter for both success and failure probably hides the reversed logic issue mentioned above - I haven't checked if that's actually the case

File: src/main/java/org/apache/sling/auth/core/impl/SlingAuthenticator.java (hunk 3 above)
Comment (anchela) [reply src/main/java/org/apache/sling/auth/core/impl/SlingAuthenticator.java:520]: i will check that....

File: src/main/java/org/apache/sling/auth/core/impl/SlingAuthenticator.java (hunk 2 above)
Comment (anchela) [reply src/main/java/org/apache/sling/auth/core/impl/SlingAuthenticator.java:519]: fine with me. i will add that.

File: src/test/java/org/apache/sling/auth/core/impl/SlingAuthenticatorOsgiTest.java (hunk 5 above)
Comment (anchela) [reply src/test/java/org/apache/sling/auth/core/impl/SlingAuthenticatorOsgiTest.java:63]: makes sense....

File: src/main/java/org/apache/sling/auth/core/impl/SlingAuthenticator.java (hunk 1 above)
Comment (anchela) [reply src/main/java/org/apache/sling/auth/core/impl/SlingAuthenticator.java:469]: was just trying to keep the style of the code, which is quite different from my coding style.
so sure.... fine by me.

File: src/main/java/org/apache/sling/auth/core/impl/SlingAuthenticationMetrics.java (hunk 4 above)
Comment (anchela) [reply src/main/java/org/apache/sling/auth/core/impl/SlingAuthenticationMetrics.java]: @bdelacretaz changed it to 'success'.

File: src/main/java/org/apache/sling/auth/core/impl/SlingAuthenticator.java (hunk 3 above)
Comment (anchela) [reply src/main/java/org/apache/sling/auth/core/impl/SlingAuthenticator.java]: you are right.... i fixed it.

**Comment by anchela** [comment]
@bdelacretaz , let me know if the changes made address your findings.

**Review by rombert — COMMENTED** [review]
Overall LGTM, but I think @cziegeler could also take a look.

The one note that I have is that now the metrics service is on the required path to set authentication up, which might impact startup time, ordering, etc. But I don't think we can avoid that in any way.

File: src/main/java/org/apache/sling/auth/core/impl/SlingAuthenticator.java [hunk 6]
```diff
@@ -457,6 +465,11 @@ private void deactivate(final BundleContext bundleContext) {
             webConsolePlugin.unregister();
             webConsolePlugin = null;
         }
+
+        metricsService = null;
+        if (metrics != null) {
```
Comment (bdelacretaz) [inline src/main/java/org/apache/sling/auth/core/impl/SlingAuthenticator.java:470]: nitpick...this test is not useful, metrics can just be set to null as no operations are called on it

**Comment by bdelacretaz** [comment]
@anchela thank your for your changes, looks good to me!

**Comment by anchela** [comment]
@bdelacretaz , @rombert thanks a lot for reviewing.
@cziegeler , would you have time to also take a look in particular wrt the comment added by @rombert , which falls in line with my original questions in the jira ticket. sling-auth-core is a pretty crucial bundle and i am bit worried by the lack of test-coverage and integration-tests and performance-benchmarks.... and i definitely would want to avoid introducing regressions... but at the same time i don't have the in depth knowledge needed to make a risk assessment nor the capacity to write all the tests that i see missing.... but i might just be overly cautious based on my experience with oak-security.

File: src/main/java/org/apache/sling/auth/core/impl/SlingAuthenticator.java [hunk 7]
```diff
@@ -363,6 +369,8 @@ private void activate(final BundleContext bundleContext,
             bundleContext, authHandlerCache);
         authInfoPostProcessorTracker = new ServiceTracker(bundleContext, AuthenticationInfoPostProcessor.class, null);
         authInfoPostProcessorTracker.open();
+
+        metrics = new SlingAuthenticationMetrics(metricsService);
     }
```
Comment (cziegeler) [inline src/main/java/org/apache/sling/auth/core/impl/SlingAuthenticator.java:374]: I suggest to pass the MetricsService in as a reference in the constructor (using constructor injection) this avoids the field for it completely

**Comment by cziegeler** [comment]
Looks ok to me, except I would prefer using constructor injection as mentioned

**Comment by cziegeler** [comment]
I'll apply the PR as is and then refactor to use constructor injection

**Comment by rombert** [comment]
> I'll apply the PR as is and then refactor to use constructor injection

@cziegeler - you could've also pushed to the branch created by the PR (not that it matters).

**Comment by cziegeler** [comment]
Good point :) will do next time, thanks

**Pull request merged**